File: windows/DotsHarness/src/DotsHarnessCore/LocalRuntimeController.cs
// Copyright (c) 2026 DOTS
// Native local model install / serve, then register as a local endpoint.

using PluginRuntime;

namespace DotsHarnessCore;

public sealed class LocalRuntimeController : ObservableObject
{
    private bool _runtimeReady;
    private string? _runtimeVersion;
    private string? _runningModelId;
    private bool _serving;
    private string _status = "Idle";
    private string? _error;
    private HashSet<string> _installed = new(StringComparer.Ordinal);
    private Dictionary<string, double> _downloads = new(StringComparer.Ordinal);

    public LocalRuntime Runtime { get; }
    public RouterController Router { get; }

    public bool RuntimeReady
    {
        get => _runtimeReady;
        private set => SetProperty(ref _runtimeReady, value);
    }

    public string? RuntimeVersion
    {
        get => _runtimeVersion;
        private set => SetProperty(ref _runtimeVersion, value);
    }

    public string? RunningModelId
    {
        get => _runningModelId;
        private set => SetProperty(ref _runningModelId, value);
    }

    public bool Serving
    {
        get => _serving;
        private set => SetProperty(ref _serving, value);
    }

    public string Status
    {
        get => _status;
        set => SetProperty(ref _status, value);
    }

    public string? Error
    {
        get => _error;
        set => SetProperty(ref _error, value);
    }

    public IReadOnlySet<string> Installed => _installed;

    public IReadOnlyDictionary<string, double> Downloads => _downloads;

    public LocalRuntimeController(SupportPaths paths, RouterController router)
    {
        Runtime = new LocalRuntime(paths);
        Router = router;
        RefreshInstalled();
        RuntimeReady = Runtime.RuntimeInstalled();
    }

    public void RefreshInstalled()
    {
        _installed = LocalModelCatalog.Models.Where(Runtime.IsInstalled).Select(m => m.Id).ToHashSet(StringComparer.Ordinal);
        Serving = Runtime.IsServing();
        RunningModelId = Runtime.RunningModelId();
        RuntimeReady = Runtime.RuntimeInstalled();
        OnPropertyChanged(nameof(Installed));
        OnPropertyChanged(nameof(Downloads));
    }

    public async Task InstallRuntimeAsync()
    {
        Error = null;
        Status = "Downloading llama.cpp…";
        SetDownload("runtime", 0);
        try
        {
            await Runtime.EnsureRuntimeAsync(progress =>
            {
                SetDownload("runtime", progress.Fraction);
                Status = $"Downloading llama.cpp {(int)(progress.Fraction * 100)}%";
            });
            RuntimeReady = true;
            SetDownload("runtime", 1);
            Status = "Local runtime ready";
        }
        catch (Exception ex)
        {
            Error = ex.Message;
            Status = "Runtime install failed";
        }
    }

    public async Task DownloadAsync(LocalModelSpec spec)
    {
        Error = null;
        SetDownload(spec.Id, 0);
        Status = $"Downloading {spec.Name}…";
        try
        {
            if (!Runtime.RuntimeInstalled()) await InstallRuntimeAsync();
            await Runtime.DownloadModelAsync(spec, progress =>
            {
                SetDownload(spec.Id, progress.Fraction);
                Status = $"{spec.Name} {(int)(progress.Fraction * 100)}%";
            });
            SetDownload(spec.Id, 1);
            RefreshInstalled();
            Status = $"{spec.Name} downloaded";
        }
        catch (Exception ex)
        {
            Error = ex.Message;
            Status = "Download failed";
        }
    }

    public async Task StartAsync(LocalModelSpec spec)
    {
        Error = null;
        try
        {
            if (!Runtime.RuntimeInstalled()) throw new RouterException("Install the local runtime first.");
            if (!Runtime.IsInstalled(spec)) throw new RouterException($"Download {spec.Name} first.");
            Status = $"Starting {spec.Name}…";
            _ = Runtime.Start(spec);
            var ready = await Runtime.WaitUntilReadyAsync();
            RefreshInstalled();
            if (!ready) throw new RouterException($"llama-server did not become ready on :{LocalRuntime.DefaultPort}");
            await PublishNodeAsync(spec);
            Status = $"Serving {spec.Name} at {Runtime.ServerUrl}";
        }
        catch (Exception ex)
        {
            Error = ex.Message;
            Status = "Start failed";
            Runtime.Stop();
            RefreshInstalled();
        }
    }

    public void Stop()
    {
        Runtime.Stop();
        RefreshInstalled();
        Status = "Local server stopped";
    }

    public void Delete(LocalModelSpec spec)
    {
        if (RunningModelId == spec.Id) Stop();
        try { File.Delete(Runtime.ModelPath(spec)); } catch { /* ignore */ }
        try { File.Delete(Runtime.ModelPath(spec) + ".part"); } catch { /* ignore */ }
        _downloads.Remove(spec.Id);
        RefreshInstalled();
        Status = $"Removed {spec.Name}";
    }

    private async Task PublishNodeAsync(LocalModelSpec spec)
    {
        Router.CustomName = spec.Name;
        Router.CustomPrefix = LocalRuntime.NodePrefix;
        Router.CustomBaseUrl = Runtime.ServerUrl.ToString().TrimEnd('/');
        Router.CustomApiKey = "";
        Router.CustomKind = CustomApiKind.OpenaiCompatible;
        Router.CustomApiType = CustomOpenAiApiType.Chat;
        var existing = Router.Nodes.FirstOrDefault(n =>
            n.Prefix == LocalRuntime.NodePrefix || n.BaseUrl.Contains($":{LocalRuntime.DefaultPort}"));
        if (existing is not null) await Router.ConnectExistingNodeAsync(existing);
        else await Router.CreateCustomNodeAsync(registerKey: true);
    }

    private void SetDownload(string id, double fraction)
    {
        _downloads[id] = fraction;
        OnPropertyChanged(nameof(Downloads));
    }
}

File: macos/DotsHarness/Sources/DotsHarnessCore/LocalRuntimeController.swift
// Copyright (c) 2026 DOTS
// Native local model install / serve, then register as a local endpoint.

import Foundation
import PluginRuntime

@MainActor
public final class LocalRuntimeController: ObservableObject {
    @Published public var runtimeReady = false
    @Published public var runtimeVersion: String?
    @Published public var runningModelID: String?
    @Published public var serving = false
    @Published public var downloads: [String: Double] = [:]
    @Published public var status: String = AppCopy.text("common.idle")
    @Published public var error: String?
    @Published public var installed: Set<String> = []

    public let runtime: LocalRuntime
    public let router: RouterController

    public init(paths: SupportPaths, router: RouterController) {
        self.runtime = LocalRuntime(paths: paths)
        self.router = router
        refreshInstalled()
        runtimeReady = runtime.runtimeInstalled()
    }

    public func refreshInstalled() {
        installed = Set(LocalModelCatalog.models.filter { runtime.isInstalled($0) }.map(\.id))
        serving = runtime.isServing()
        runningModelID = runtime.runningModelID()
        runtimeReady = runtime.runtimeInstalled()
    }

    public func installRuntime() async {
        error = nil
        status = AppCopy.text("local.downloadingRuntime")
        downloads["runtime"] = 0
        do {
            try await runtime.ensureRuntime { [weak self] progress in
                Task { @MainActor in
                    self?.downloads["runtime"] = progress.fraction
                    self?.status = AppCopy.format("local.downloadingRuntimeProgress", Int(progress.fraction * 100))
                }
            }
            runtimeReady = true
            downloads["runtime"] = 1
            status = AppCopy.text("local.runtimeReady")
        } catch {
            self.error = error.localizedDescription
            status = AppCopy.text("local.runtimeInstallFailed")
        }
    }

    public func download(_ spec: LocalModelSpec) async {
        error = nil
        downloads[spec.id] = 0
        status = AppCopy.format("local.downloadingModel", spec.name)
        do {
            if !runtime.runtimeInstalled() {
                await installRuntime()
            }
            try await runtime.downloadModel(spec) { [weak self] progress in
                Task { @MainActor in
                    self?.downloads[spec.id] = progress.fraction
                    self?.status = AppCopy.format("local.modelProgress", spec.name, Int(progress.fraction * 100))
                }
            }
            downloads[spec.id] = 1
            refreshInstalled()
            status = AppCopy.format("local.modelDownloaded", spec.name)
        } catch {
            self.error = error.localizedDescription
            status = AppCopy.text("local.downloadFailed")
        }
    }

    public func start(_ spec: LocalModelSpec) async {
        error = nil
        do {
            guard runtime.runtimeInstalled() else {
                throw RouterError(AppCopy.text("localRuntime.installFirst"))
            }
            guard runtime.isInstalled(spec) else {
                throw RouterError(AppCopy.format("localRuntime.downloadFirst", spec.name))
            }
            status = AppCopy.format("local.startingModel", spec.name)
            _ = try runtime.start(model: spec)
            let ready = await runtime.waitUntilReady()
            refreshInstalled()
            if !ready {
                throw RouterError(AppCopy.format("local.serverNotReady", LocalRuntime.defaultPort))
            }
            await publishNode(for: spec)
            status = AppCopy.format("local.servingModel", spec.name, runtime.serverURL.absoluteString)
        } catch {
            self.error = error.localizedDescription
            status = AppCopy.text("local.startFailed")
            runtime.stop()
            refreshInstalled()
        }
    }

    public func stop() {
        runtime.stop()
        refreshInstalled()
        status = AppCopy.text("local.serverStopped")
    }

    public func delete(_ spec: LocalModelSpec) {
        if runningModelID == spec.id { stop() }
        try? FileManager.default.removeItem(at: runtime.modelURL(spec))
        try? FileManager.default.removeItem(at: runtime.modelURL(spec).appendingPathExtension("part"))
        downloads[spec.id] = nil
        refreshInstalled()
        status = AppCopy.format("local.modelRemoved", spec.name)
    }

    private func publishNode(for spec: LocalModelSpec) async {
        router.customName = spec.name
        router.customPrefix = LocalRuntime.nodePrefix
        router.customBaseURL = runtime.serverURL.absoluteString
        router.customAPIKey = ""
        router.customKind = .openai
        router.customAPIType = .chat
        if let existing = router.nodes.first(where: { $0.prefix == LocalRuntime.nodePrefix || $0.baseURL.contains(":\(LocalRuntime.defaultPort)") }) {
            await router.connectExistingNode(existing)
        } else {
            await router.createCustomNode(registerKey: true)
        }
    }
}
